### PR TITLE
Add debug symbols support #9792

### LINKED ISSUE #9792: Offer quarkus property to enable debug symbols
Now that GraalVM has debug symbols, have a Quarkus property

e.g. `quarkus.native.debug-symbols`

so it generates the image with debug symbols.

Useful in conjunction with profiles

### PR DESCRIPTION
**Pull request by galderz**
Closes #9792. 

This PR adds:
* A new debug symbol enable option called `quarkus.native.debug.enabled`.
* A new option has been added instead of breaking existing users that might rely on `quarkus.native.debugSymbols` flag. This flag adds `-g` to the `native-image` invocation which is a GraalVM EE only flag. Should it eventually deprecated? 
* After creating the native image, if `objcopy` is found, it uses it to split the debug symbols from the binary executable. The debug symbols are put in a file called `<executable>.debug` at the same level as the binary.
* This new option is disabled by default. See [quarkus-dev](https://groups.google.com/forum/#!topic/quarkus-dev/VOSQ6zhy_vI) discussion for details.

Q. What can I debug with this?
* User code in Quarkus project.
* JDK code.
* Some of the `com.oracle.svm` code.

Q. What can't I debug?
* Library dependencies. Further work on Quarkus build code needs to be done to enable source jars to be found. ~I'll create a follow up issue for this~. See #10094.
* `org.graalvm` code. Some sources are missing from latest GraalVM CE downloads. @adinn already spotted this and reported it in https://github.com/oracle/graal/issues/2551.

Q. What I do need to do native debugging?
* `gdb` version 8.x or 9.x. I tried with 7.6 but gdb itself gave me a segmentation fault (working with @adinn to figure out what's wrong there).
* (optional) emacs so that you can have nice integration between gdb and source navigation.

Q. Can I give this a try? 
Sure! 

I've been using [this sample project called `debug-info`](https://github.com/galderz/debug-info) to test things out. This is how it works:

1. Build this Quarkus branch locally.
2. Go into debug-info project and call (assumes you're using GraalVM CE 20.1.0 as `JAVA_HOME`):
```
./mvnw clean package -DskipTests \
	-Pnative \
	-Dquarkus.platform.version=999-SNAPSHOT \
	-Dquarkus-plugin.version=999-SNAPSHOT \
	-Dquarkus.platform.artifact-id=quarkus-bom \
	-Dquarkus.native.debug.enabled=true
```
3. Fire up `emacs` (or any other tool that has gdb integration) from the root of the project.
4. Call `M-x gud-gdb`
5. Set directories for source lookup: `set directories ../src/main/java:debug-info-1.0.0-SNAPSHOT-native-image-source-jar/sources/graal:debug-info-1.0.0-SNAPSHOT-native-image-source-jar/sources/jdk`
6. Invoke `gdb`: `gdb --fullname target/debug-info-1.0.0-SNAPSHOT-runner`
7. Verify that debug symbols are present, e.g. `info func greeting`
8. Put a breakpoint in the code, e.g. `break debug.info.GreetingService::greeting`
9. Make `gdb` execute the program, e.g. `run`
10. From a separate shell exercise the app, e.g. `curl localhost:8080/hello/greeting/quarkus`. The debugger should stop in the breakpoint.
11. Type `s` or `n` to jump around the code.

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java [hunk 1]
```diff
@@ -528,4 +535,48 @@ private static String testGCCArgument(String argument) {
...
+        }
+
+        log.info("Cannot find executable (objcopy) to separate symbols from executable.");
+        return false;
+    }
+
+    private void splitDebugSymbols(Path executable) {
+        Path symbols = Paths.get(String.format("%s.debug", executable.toString()));
+        objcopy("--only-keep-debug", executable.toString(), symbols.toString());
+        objcopy("--strip-debug", executable.toString());
```
Comment (jerboaa) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:561]: I'd suggest to run this `objcopy("--strip-debug", executable.toString());` also when no debug symbols are being requested.  The reason for this is that OpenJDK's static libraries might contain debug symbols which in turn might end up in the native image binary and you could get rid of them in the produced native image by stripping them via `objcopy --strip-debug`.

**Review by jerboaa — COMMENTED** [review]
Have you considered also to run the `--strip-debug` option on the produced image even when `-H:GenerateDebugInfo=0`? This is not for the application's debug symbols to get stripped, but for potential debug symbols in linked in static libraries to get removed. Smaller binaries and all that ;-)

**Comment by galderz** [comment]
@jerboaa Updated the PR with your suggestion. Can you have another look? I assumed it's fine to strip debug symbols after adding the gnu debug link?

**Comment by jerboaa** [comment]
> @jerboaa Updated the PR with your suggestion. Can you have another look? I assumed it's fine to strip debug symbols after adding the gnu debug link?

This seems fine to me. Thanks for the update.

**Comment by jaikiran** [comment]
Just curious - is this planned to be available for 1.6 release? Would be a good thing to have, I think.

**Review by gsmet — CHANGES_REQUESTED** [review]
Spotted a small typo and forgot to post my review yesterday :/

**Comment by gsmet** [comment]
@galderz where are we on this? Did anyone agree with this approach or are there still ongoing discussions?

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java [hunk 2]
```diff
@@ -359,6 +359,14 @@ public NativeImageBuildItem build(NativeConfig nativeConfig, NativeImageSourceJa
             Files.delete(generatedImage);
             System.setProperty("native.image.path", finalPath.toAbsolutePath().toString());
 
+            if (objcopyExists(env)) {
+                if (nativeConfig.debug.enabled) {
+                    splitDebugSymbols(finalPath);
+                }
+                // Strip debug symbols regardless, because the the underlying JDK might contain them
```
[the remark this answers is not in the document]
Comment (gsmet) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java]: I just pushed a fix.

**Comment by gsmet** [comment]
@galderz can you confirm me that we should merge this?

**Review by gsmet — APPROVED** [review]

**Comment by galderz** [comment]
@gsmet All looks good after [this discussion](https://groups.google.com/g/quarkus-dev/c/VOSQ6zhy_vI/m/56xzfbTMBwAJ). I'm working on further enhancements that cannot make it this release (e.g. support third party library source code navigation).

**Comment by gsmet** [comment]
Ok, let's merge it then, thanks!

**Pull request merged**